File: Backend/Travle.Model/Responses/RoleApplicationResponse.cs
namespace Travle.Model.Responses
{
    /// <summary>
    /// A user's request for the Curator or Organizer role, with its full decision history. The
    /// supporting <c>Document</c> bytes are never carried here (§8.2 — no heavy payloads in lists);
    /// <see cref="HasDocument"/> tells the client whether to offer the dedicated download endpoint.
    /// </summary>
    public class RoleApplicationResponse
    {
        public int Id { get; set; }

        public int UserId { get; set; }
        /// <summary>Flattened from User.Username so the moderation queue never shows a raw id.</summary>
        public string? Username { get; set; }
        /// <summary>Flattened "FirstName LastName" of the applicant.</summary>
        public string? ApplicantName { get; set; }

        public int RoleId { get; set; }
        /// <summary>The applied-for role name (Curator or Organizer), flattened from Role.Name.</summary>
        public string? RoleName { get; set; }

        public string Motivation { get; set; } = string.Empty;

        public int? RegionId { get; set; }
        public string? RegionName { get; set; }

        /// <summary>Whether a supporting document is attached (fetch it via the document endpoint).</summary>
        public bool HasDocument { get; set; }
        public string? DocumentContentType { get; set; }

        /// <summary>Pending / Approved / Rejected — the enum name, never the raw int.</summary>
        public string Status { get; set; } = string.Empty;

        public int? DecidedByUserId { get; set; }
        public string? DecidedByUsername { get; set; }
        public DateTime? DecidedAt { get; set; }
        public string? RejectionReason { get; set; }

        public DateTime CreatedAt { get; set; }
        public DateTime? ModifiedAt { get; set; }
    }
}

File: UI/travle_core/lib/src/models/role_application_response.dart
import 'package:json_annotation/json_annotation.dart';

part 'role_application_response.g.dart';

/// A user's Curator/Organizer role application with its decision history (mirrors
/// the backend `RoleApplicationResponse`). The document bytes are never carried
/// here — [hasDocument] tells the client whether to offer the download endpoint.
@JsonSerializable()
class RoleApplicationResponse {
  RoleApplicationResponse({
    required this.id,
    required this.userId,
    required this.roleId,
    required this.motivation,
    required this.hasDocument,
    required this.status,
    required this.createdAt,
    this.username,
    this.applicantName,
    this.roleName,
    this.regionId,
    this.regionName,
    this.documentContentType,
    this.decidedByUserId,
    this.decidedByUsername,
    this.decidedAt,
    this.rejectionReason,
    this.modifiedAt,
  });

  final int id;
  final int userId;
  final String? username;
  final String? applicantName;
  final int roleId;
  final String? roleName;
  final String motivation;
  final int? regionId;
  final String? regionName;
  final bool hasDocument;
  final String? documentContentType;

  /// Pending / Approved / Rejected (the enum name, never a raw int).
  final String status;

  final int? decidedByUserId;
  final String? decidedByUsername;
  final DateTime? decidedAt;
  final String? rejectionReason;
  final DateTime createdAt;
  final DateTime? modifiedAt;

  bool get isPending => status == 'Pending';
  bool get isApproved => status == 'Approved';
  bool get isRejected => status == 'Rejected';

  factory RoleApplicationResponse.fromJson(Map<String, dynamic> json) =>
      _$RoleApplicationResponseFromJson(json);

  Map<String, dynamic> toJson() => _$RoleApplicationResponseToJson(this);
}
